File: refer/volcengine_unidirectional_stream_demo(python)/examples/volcengine/unidirectional_stream.py
#!/usr/bin/env python3
import argparse
import json
import logging
import uuid

import websockets

from protocols import EventType, MsgType, full_client_request, receive_message

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def get_resource_id(voice: str) -> str:
    if voice.startswith("S_"):
        return "volc.megatts.default"
    return "volc.service_type.10029"


async def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--appid", required=True, help="APP ID")
    parser.add_argument("--access_token", required=True, help="Access Token")
    parser.add_argument("--resource-id", default="", help="Resource ID")
    parser.add_argument("--text", required=True, help="Text to convert")
    parser.add_argument("--voice_type", default="", required=True, help="Voice type")
    parser.add_argument("--encoding", default="wav", help="Output file encoding")
    parser.add_argument(
        "--endpoint",
        default="wss://openspeech.bytedance.com/api/v3/tts/unidirectional/stream",
        help="WebSocket endpoint URL",
    )

    args = parser.parse_args()

    # Connect to server
    headers = {
        "X-Api-App-Key": args.appid,
        "X-Api-Access-Key": args.access_token,
        "X-Api-Resource-Id": (
            args.resource_id if args.resource_id else get_resource_id(args.voice_type)
        ),
        "X-Api-Connect-Id": str(uuid.uuid4()),
    }

    logger.info(f"Connecting to {args.endpoint} with headers: {headers}")
    websocket = await websockets.connect(
        args.endpoint, additional_headers=headers, max_size=10 * 1024 * 1024
    )
    logger.info(
        f"Connected to WebSocket server, Logid: {websocket.response.headers['x-tt-logid']}",
    )

    try:
        # Prepare request payload
        request = {
            "user": {
                "uid": str(uuid.uuid4()),
            },
            "req_params": {
                "speaker": args.voice_type,
                "audio_params": {
                    "format": args.encoding,
                    "sample_rate": 24000,
                    "enable_timestamp": True,
                },
                "text": args.text,
                "additions": json.dumps(
                    {
                        "disable_markdown_filter": False,
                    }
                ),
            },
        }

        # Send request
        await full_client_request(websocket, json.dumps(request).encode())

        # Receive audio data
        audio_data = bytearray()
        while True:
            msg = await receive_message(websocket)

            if msg.type == MsgType.FullServerResponse:
                if msg.event == EventType.SessionFinished:
                    break
            elif msg.type == MsgType.AudioOnlyServer:
                audio_data.extend(msg.payload)
            else:
                raise RuntimeError(f"TTS conversion failed: {msg}")

        # Check if we received any audio data
        if not audio_data:
            raise RuntimeError("No audio data received")

        # Save audio file
        filename = f"{args.voice_type}.{args.encoding}"
        with open(filename, "wb") as f:
            f.write(audio_data)
        logger.info(f"Audio received: {len(audio_data)}, saved to {filename}")

    finally:
        await websocket.close()
        logger.info("Connection closed")


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())

File: refer/volcengine_unidirectional_stream_demo.tar(java)/volcengine_unidirectional_stream_demo(java)/src/main/java/com/speech/volcengine/UnidirectionalStream.java
package com.speech.volcengine;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.speech.protocol.EventType;
import com.speech.protocol.Message;
import com.speech.protocol.MsgType;
import com.speech.protocol.SpeechWebSocketClient;
import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.io.ByteArrayOutputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.Map;
import java.util.UUID;

@Slf4j
public class UnidirectionalStream {
    private static final String ENDPOINT = "wss://openspeech.bytedance.com/api/v3/tts/unidirectional/stream";
    private static final ObjectMapper objectMapper = new ObjectMapper();

    /**
     * Get resource ID based on voice type
     *
     * @param voice Voice type string
     * @return Corresponding resource ID
     */
    public static String voiceToResourceId(String voice) {
        // Map different voice types to resource IDs based on actual needs
        if (voice.startsWith("S_")) {
            return "volc.megatts.default";
        }
        return "volc.service_type.10029";
    }

    public static void main(String[] args) throws Exception {
        // Configure parameters
        String appId = System.getProperty("appId");
        String accessToken = System.getProperty("accessToken");
        String voice = System.getProperty("voice", "");
        String text = System.getProperty("text", "");
        String encoding = System.getProperty("encoding", "wav");

        if (appId == null || accessToken == null) {
            log.error("Please set appId and accessToken system properties");
            System.exit(1);
        }

        // Set request headers
        Map<String, String> headers = Map.of(
                "X-Api-App-Key", appId,
                "X-Api-Access-Key", accessToken,
                "X-Api-Resource-Id", voiceToResourceId(voice),
                "X-Api-Connect-Id", UUID.randomUUID().toString());

        // Create WebSocket client
        SpeechWebSocketClient client = new SpeechWebSocketClient(new URI(ENDPOINT), headers);
        try {
            client.connectBlocking();
            // Prepare request parameters
            Map<String, Object> request = Map.of(
                    "user", Map.of("uid", UUID.randomUUID().toString()),
                    "req_params", Map.of(
                            "speaker", voice,
                            "audio_params", Map.of(
                                    "format", encoding,
                                    "sample_rate", 24000,
                                    "enable_timestamp", true),
                            // additions requires a JSON string
                            "additions", objectMapper.writeValueAsString(Map.of(
                                    "disable_markdown_filter", false)),
                            "text", text));

            // Send request
            client.sendFullClientMessage(objectMapper.writeValueAsBytes(request));

            // Receive response
            ByteArrayOutputStream audioStream = new ByteArrayOutputStream();
            while (true) {
                Message msg = client.receiveMessage();
                log.info("Received message: {}", msg);

                if (msg.getType() == MsgType.AUDIO_ONLY_SERVER) {
                    if (msg.getPayload() != null) {
                        audioStream.write(msg.getPayload());
                    }
                } else if (msg.getType() == MsgType.ERROR) {
                    throw new RuntimeException("Server returned error: " + new String(msg.getPayload()));
                }

                if (msg.getType() == MsgType.FULL_SERVER_RESPONSE &&
                        msg.getEvent() == EventType.TTS_SENTENCE_END) {
                    String jsonString = new String(msg.getPayload(), StandardCharsets.UTF_8);
                    log.info("Received TTS response sentence end: {}", jsonString);
                    continue;
                }

                if (msg.getType() == MsgType.FULL_SERVER_RESPONSE &&
                        msg.getEvent() == EventType.SESSION_FINISHED) {
                    break;
                }
            }

            if (audioStream.size() == 0) {
                throw new RuntimeException("No audio data received");
            }

            // Save audio file
            String fileName = String.format("%s.%s", voice, encoding);
            Files.write(new File(fileName).toPath(), audioStream.toByteArray());
            log.info("Audio saved to file: {}", fileName);
        } finally {
            client.closeBlocking();
        }
    }
}